### ISSUE TITLE
Pods mounting wrong PV during CSI migration on Azure

### ISSUE THREAD

**Issue by kon-angelo** [kubernetes/kubernetes#111006]
### What happened?

We tried to upgrade our cluster to use CSI on kubernetes v1.21.12. During the migration process we deploy the CSI components and start creating new nodes with the appropriate settings while draining and decommissioning old nodes.

During the transitioning phase where both CSI and non-CSI nodes exist in the cluster, we noticed that some of our workload that uses azuredisk PVCs were mounting the wrong Volumes. To be more precise, we noticed that certain pods were writing to the volume that belonged to another pod that was scheduled on the same node. Given that these are RWO volumes, this is a behaviour that should not happen.

This PVC "mangling" happens because we can manage to have 2 PVCs use the same LUN on an Azure VM. Before outlining the incident timeline, an important information is that our workload is a deployment that mounts a pre-provisioned PVC. The incident timeline is the following:

1) The pod (pod `A1`) referencing PVC `A` on the old node is set to terminating. Because it belongs to a deployment, a new pod (pod `A2`) referencing  the same PVC (`A`) is immediately scheduled in the new CSI-enabled node.
2) `volumeattachment` is created for the new node and azuredisk-csi tries to attach the volume. It gets a dangling volume error because `A1` is still in `Terminating` state. It then forcefully detaches the volume and attaches it to the new node in `LUN 0` all the while pod `A1` is still up. Pod `A2` is now running.
3) kube-controller-manager reconciles its desired-state-of-the world. Because the `A1` is scheduled in a non-CSI node, KCM uses the in-tree provider to reconcile the volume. It sees that `A1` exists still (even in `Terminating`) but because the  volume is not attached to the node, it forcefully detaches the volume from the CSI node and puts it back to the old one. `LUN 0` in the new node is now free.

Now, the new pod is still running even though its volume has already been detached. This is already a problematic state because a stateful workload may not have enough time to properly terminate but the disk controller will not respect that. The check that the volume is only mounted once (because the volume is a `RWO` one) is also not respected during the transition. 

4) A new workload (pod `B`) using another PVC (PVC `B`) is scheduled to the new CSI-enabled node. Because `LUN 0` is now free, azuredisk-csi will use it (as the first available one) for the new workload. 

At this point, we have 2 `volumeattachments` referencing the same LUN. However only volume `B` is attached to the node at that time. 

5?) Even thought we have the `volumeattachments` using the same LUN, its not enough for both pods to have access to the volume. Pod `A2` is still running, but because of the detaching of its volume, the mountpoint inside its container is broken as well as the `globalmountpoint` for the volume `A` of pod `A2`. Restarting the kubelet was enough to repair the `globalmountpoint` (I think as part of the [azuredisk-csi `NodePublishVolume`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/751203d138906cbf1f4de5c020a44c7e56de3c12/pkg/azuredisk/nodeserver.go#L581)). Restarting the container of `A2` after that would give it access to volume `B`.

/sig storage

### What did you expect to happen?

Each pod to mount the correct volume. Even more important, was not to have multiple pods writing to the same RWO volume.

### How can we reproduce it (as minimally and precisely as possible)?

The following steps can be used to reproduce the incident, annotated with 
+ Create a PVC
+ Create a deployment that uses the PVC. The deployment should emulate a pod that stays terminating for some time. For this step I used a simple deployment that ignores SIGTERM and has a very high `terminationGracePeriodSeconds`. 
+ After that, start performing the upgrade process to use CSI.
+ Set the old pod to terminating
+ Wait until the new pod is running
+ Restart kube-controller-manager and wait until the volume is detached from the new node.
+ Create a new stateful workload on the CSI node (anything with PVC).
+ Restart kubelet
+ Restart container

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
v1.21.12
</details>

### Cloud provider

<details>
Azure
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
azuredisk-csi v1.16.0
</details>

**Comment by andyzhangx** [comment]
hi @kon-angelo what does `non-CSI node` mean?

**Comment by kon-angelo** [comment]
Hello @andyzhangx. When we enable the CSI for a running cluster, we gradually introduce nodes with CSI enabled while we phase the old ones using the in-tree code. 

During the transition period, the API Server is running with `CSIMigrationAzureDisk=true` featuregate. However only after every node is replaced we enable the  `InTreePluginAzureDiskUnregister=true` featuregate. 

For your question, the csi-driver daemonset will be running on all nodes, however the kubelet of old nodes is still using the in-tree code. Once all "old" nodes are  replaced and `InTreePluginAzureDiskUnregister=true` is used, then we don't see this issue anymore. 

The core of the issue is that csi-attacher and KCM-attacher are competing controllers that:
1) during the transition period both are active
2) operate on the same infrastructure object (azure disks)
3) but watch different k8s objects (volumeattachments vs pods) 

AFAIK what makes Azure "special" compared to the same setup in other providers, is that the Azure controllers can **force** detach the volumes, hence you end up in a situation where a volume may get ping-pong'd between CSI and "in-tree" nodes.

**Comment by andyzhangx** [comment]
have you set `CSIMigration=true,CSIMigrationAzureDisk=true` for kube-controller-manager? if it's set, all requests would be handled by csi driver, kcm won't do disk attach/detach.

**Comment by kon-angelo** [comment]
@andyzhangx yes they were set. The logs of the KCM that I saw though had attach/detach messages printed.

**Comment by andyzhangx** [comment]
but if old pod is in terminating state, it should tries to detach the disk, why would it attach the disk again? still confused for me.

**Comment by kon-angelo** [comment]
True, this is a weird behavior. The pod is terminating but the container is still running (ignoring SIGTERM). The volume is still in the node's attach and in-use list and I assume that KCM reconciles based on that information. If it did not try to reattach the volume, then this issue wouldn't happen.
